File: rewrite-core/src/main/java/com/netflix/rewrite/Refactor.java
/*
 * Copyright 2020 the original authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite;

import com.netflix.rewrite.internal.lang.NonNullApi;
import com.netflix.rewrite.internal.lang.Nullable;
import com.netflix.rewrite.tree.*;
import com.netflix.rewrite.visitor.refactor.RefactorVisitor;
import com.netflix.rewrite.visitor.refactor.TransformVisitor;
import com.netflix.rewrite.visitor.refactor.op.*;
import com.netflix.rewrite.visitor.refactor.op.*;

import java.util.*;
import java.util.function.Function;

import static com.netflix.rewrite.tree.Formatting.format;
import static com.netflix.rewrite.tree.Tr.randomId;

@NonNullApi
public class Refactor {
    private final Tr.CompilationUnit original;

    private final List<RefactorVisitor> ops = new ArrayList<>();

    public Refactor(Tr.CompilationUnit original) {
        this.original = original;
    }

    // -------------
    // Custom refactoring visitors
    // -------------

    public Refactor visit(Iterable<RefactorVisitor> visitors) {
        visitors.forEach(ops::add);
        return this;
    }

    public Refactor visit(RefactorVisitor visitor) {
        ops.add(visitor);
        return this;
    }

    // -------------
    // Compilation Unit Refactoring
    // -------------

    public Refactor addImport(String clazz) {
        return addImport(clazz, null);
    }

    public Refactor addImport(String clazz, @Nullable String staticMethod) {
        return addImport(clazz, staticMethod, false);
    }

    public Refactor addImport(String clazz, @Nullable String staticMethod, boolean onlyIfReferenced) {
        ops.add(new AddImport(clazz, staticMethod, onlyIfReferenced));
        return this;
    }

    public Refactor removeImport(String clazz) {
        ops.add(new RemoveImport(clazz));
        return this;
    }

    // -------------
    // Class Declaration Refactoring
    // -------------

    public Refactor addField(Tr.ClassDecl target, String clazz, String name) {
        return addField(target, clazz, name, null);
    }

    public Refactor addField(Tr.ClassDecl target, String clazz, String name, @Nullable String init) {
        ops.add(new AddField(target.getId(), Collections.singletonList(new Tr.Modifier.Private(randomId(), format("", " "))), clazz, name, init));
        ops.add(new AddImport(clazz, null, false));
        return this;
    }

    // -------------
    // Field Refactoring
    // -------------

    public Refactor changeFieldType(Iterable<Tr.VariableDecls> targets, String toType) {
        targets.forEach(t -> changeFieldType(t, toType));
        return this;
    }

    public Refactor changeFieldType(Tr.VariableDecls target, String toType) {
        ops.add(new ChangeFieldType(target.getId(), toType));
        return this;
    }

    public Refactor changeFieldName(Type.Class classType, String hasName, String toName) {
        ops.add(new ChangeFieldName(classType, hasName, toName));
        return this;
    }

    public Refactor deleteStatement(Iterable<Statement> targets) {
        targets.forEach(this::deleteStatement);
        return this;
    }

    public Refactor deleteStatement(Statement statement) {
        ops.add(new DeleteStatement(statement.getId()));
        return this;
    }

    // -------------
    // Method Refactoring
    // -------------

    public Refactor changeMethodName(Iterable<Tr.MethodInvocation> targets, String name) {
        targets.forEach(t -> changeMethodName(t, name));
        return this;
    }

    public Refactor changeMethodName(Tr.MethodInvocation target, String toName) {
        ops.add(new ChangeMethodName(target.getId(), toName));
        return this;
    }

    public Refactor changeMethodTargetToStatic(Iterable<Tr.MethodInvocation> targets, String toClass) {
        targets.forEach(t -> changeMethodTargetToStatic(t, toClass));
        return this;
    }

    public Refactor changeMethodTargetToStatic(Tr.MethodInvocation target, String toClass) {
        ops.add(new ChangeMethodTargetToStatic(target.getId(), toClass));
        return this;
    }

    public Refactor changeMethodTarget(Iterable<Tr.MethodInvocation> targets, Tr.VariableDecls.NamedVar namedVar) {
        targets.forEach(t -> changeMethodTarget(t, namedVar));
        return this;
    }

    public Refactor changeMethodTarget(Tr.MethodInvocation target, Tr.VariableDecls.NamedVar namedVar) {
        return changeMethodTarget(target, namedVar.getSimpleName(), TypeUtils.asClass(namedVar.getType()));
    }

    public Refactor changeMethodTarget(Iterable<Tr.MethodInvocation> targets, String namedVar, @Nullable Type.Class clazz) {
        targets.forEach(t -> changeMethodTarget(t, namedVar, clazz));
        return this;
    }

    public Refactor changeMethodTarget(Tr.MethodInvocation target, String namedVar, @Nullable Type.Class clazz) {
        ops.add(new ChangeMethodTargetToVariable(target.getId(), namedVar, clazz));
        return this;
    }

    public Refactor insertArgument(Iterable<Tr.MethodInvocation> targets, int pos, String source) {
        targets.forEach(t -> insertArgument(t, pos, source));
        return this;
    }

    public Refactor insertArgument(Tr.MethodInvocation target, int pos, String source) {
        ops.add(new InsertMethodArgument(target.getId(), pos, source));
        return this;
    }

    public Refactor deleteArgument(Iterable<Tr.MethodInvocation> targets, int pos) {
        targets.forEach(t -> deleteArgument(t, pos));
        return this;
    }

    public Refactor deleteArgument(Tr.MethodInvocation target, int pos) {
        ops.add(new DeleteMethodArgument(target.getId(), pos));
        return this;
    }

    public ReorderMethodArguments reorderArguments(Tr.MethodInvocation target, String... byArgumentNames) {
        ReorderMethodArguments reorder = new ReorderMethodArguments(target.getId(), byArgumentNames);
        ops.add(reorder);
        return reorder;
    }

    // -------------
    // Expression Refactoring
    // -------------

    public Refactor changeLiteral(Iterable<Expression> targets, Function<Object, Object> transform) {
        targets.forEach(t -> changeLiteral(t, transform));
        return this;
    }

    public Refactor changeLiteral(Expression target, Function<Object, Object> transform) {
        ops.add(new ChangeLiteral(target.getId(), transform));
        return this;
    }

    public Refactor changeType(String from, String to) {
        ops.add(new ChangeType(from, to));
        return this;
    }

    public RefactorResult fix() {
        return fix(10);
    }

    public RefactorResult fix(int maxCycles) {
        Tr.CompilationUnit acc = original;
        Set<String> rulesThatMadeChanges = new HashSet<>();

        for (int i = 0; i < maxCycles; i++) {
            Set<String> rulesThatMadeChangesThisCycle = new HashSet<>();
            for (RefactorVisitor visitor : ops) {
                // only for use in debugging visitors
                visitor.setCycle(i);

                if(visitor.isSingleRun() && i > 0) {
                    continue;
                }

                var before = acc;
                acc = transformRecursive(acc, visitor);
                if(before != acc) {
                    // we only report on the top-level visitors, not any andThen() visitors that
                    // are applied as part of the top-level visitor's pipeline
                    rulesThatMadeChangesThisCycle.add(visitor.getRuleName());
                }
            }
            if (rulesThatMadeChangesThisCycle.isEmpty()) {
                break;
            }
            rulesThatMadeChanges.addAll(rulesThatMadeChangesThisCycle);
        }

        return new RefactorResult(original, acc, rulesThatMadeChanges);
    }

    private Tr.CompilationUnit transformRecursive(Tr.CompilationUnit acc, RefactorVisitor visitor) {
        // by transforming the AST for each op, we allow for the possibility of overlapping changes
        acc = (Tr.CompilationUnit) new TransformVisitor(visitor.visit(acc)).visit(acc);
        for (RefactorVisitor vis : visitor.andThen()) {
            acc = transformRecursive(acc, vis);
        }
        return acc;
    }
}

File: rewrite/src/main/kotlin/com/netflix/rewrite/refactor/Refactor.kt
/**
 * Copyright 2016 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.refactor

import com.netflix.rewrite.ast.*
import com.netflix.rewrite.ast.visitor.FormatVisitor
import com.netflix.rewrite.ast.visitor.RetrieveTreeVisitor
import com.netflix.rewrite.ast.visitor.TransformVisitor
import com.netflix.rewrite.refactor.op.*
import org.eclipse.jgit.diff.DiffEntry
import org.eclipse.jgit.diff.DiffFormatter
import org.eclipse.jgit.internal.storage.dfs.InMemoryRepository
import org.eclipse.jgit.lib.Constants
import org.eclipse.jgit.lib.FileMode
import java.io.ByteArrayOutputStream
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*

open class Refactor(val original: Tr.CompilationUnit) {
    /**
     * The operation's target may change if another operation transforms the original target first
     */
    private data class RefactorOperation(var id: Long, val visitor: RefactorVisitor<*>)
    private val ops = ArrayList<RefactorOperation>()

    private fun addOp(target: Tree, visitor: RefactorVisitor<*>) {
        ops.add(RefactorOperation(target.id, visitor))
    }

    private fun addOp(visitor: RefactorVisitor<*>) {
        ops.add(RefactorOperation(original.id, visitor))
    }

    // -------------
    // Compilation Unit Refactoring
    // -------------

    @JvmOverloads
    fun addImport(clazz: Class<*>, staticMethod: String? = null): Refactor =
        addImport(clazz.name, staticMethod)

    @JvmOverloads
    fun addImport(clazz: String, staticMethod: String? = null): Refactor {
        addOp(AddImport(clazz, staticMethod))
        return this
    }

    fun removeImport(clazz: Class<*>): Refactor = removeImport(clazz.name)

    fun removeImport(clazz: String): Refactor {
        addOp(RemoveImport(clazz))
        return this
    }

    fun changeType(from: Class<*>, to: Class<*>): Refactor = changeType(from.name, to.name)

    fun changeType(from: String, to: String): Refactor {
        addOp(ChangeType(from, to))
        addOp(AddImport(to, onlyIfReferenced = true))
        addOp(RemoveImport(from))
        return this
    }

    fun run(t: Tree, visitor: RefactorVisitor<*>): Refactor {
        addOp(t, visitor)
        return this
    }

    fun run(visitor: RefactorVisitor<*>): Refactor {
        addOp(visitor)
        return this
    }

    // -------------
    // Class Declaration Refactoring
    // -------------

    @JvmOverloads
    fun addField(target: Tr.ClassDecl, clazz: Class<*>, name: String, init: String? = null) =
        addField(target, clazz.name, name, init)

    @JvmOverloads
    fun addField(target: Tr.ClassDecl, clazz: String, name: String, init: String? = null): Refactor {
        addOp(target, AddField(listOf(Tr.Modifier.Private(format("", " "))), clazz, name, init))
        addOp(AddImport(clazz))
        return this
    }

    // -------------
    // Field Refactoring
    // -------------

    fun changeFieldType(targets: Iterable<Tr.VariableDecls>, toType: String): Refactor {
        targets.forEach { target ->
            addOp(target, ChangeFieldType(toType))
            target.typeExpr.type?.asClass()?.let { addOp(RemoveImport(it.fullyQualifiedName)) }
        }

        if(targets.any())
            addOp(AddImport(toType))

        return this
    }

    fun changeFieldType(target: Tr.VariableDecls, toType: Class<*>) =
            changeFieldType(target, toType.name)

    fun changeFieldType(target: Tr.VariableDecls, toType: String) =
            changeFieldType(listOf(target), toType)

    fun changeFieldName(targets: Iterable<Tr.VariableDecls>, toName: String): Refactor {
        targets.forEach { target -> addOp(target, ChangeFieldName(toName)) }
        return this
    }

    fun changeFieldName(target: Tr.VariableDecls, toName: String) =
            changeFieldName(listOf(target), toName)

    fun deleteField(targets: Iterable<Tr.VariableDecls>): Refactor {
        targets.groupBy { original.cursor(it.id)!!.enclosingClass() }
                .forEach { clazz, variables ->
                    addOp(clazz!!, DeleteField(variables))
                    variables.forEach { (_, _, typeExpr) ->
                        typeExpr.type?.asClass()?.let { addOp(RemoveImport(it.fullyQualifiedName)) }
                    }
                }

        return this
    }

    fun deleteField(target: Tr.VariableDecls) = deleteField(listOf(target))

    // -------------
    // Method Refactoring
    // -------------

    fun changeMethodName(targets: Iterable<Tr.MethodInvocation>, toName: String): Refactor {
        targets.forEach { target -> addOp(target, ChangeMethodName(toName)) }
        return this
    }

    fun changeMethodName(target: Tr.MethodInvocation, toName: String) =
            changeMethodName(listOf(target), toName)

    /**
     * Change to a static method invocation on <code>toClass</code>
     */
    fun changeMethodTargetToStatic(targets: Iterable<Tr.MethodInvocation>, toClass: String): Refactor {
        targets.forEach { target ->
            addOp(target, ChangeMethodTargetToStatic(toClass))
            target.type?.declaringType?.fullyQualifiedName?.let { addOp(RemoveImport(it)) }
        }

        if(targets.any())
            addOp(AddImport(toClass))

        return this
    }

    /**
     * Change to a static method invocation on <code> toClass
     */
    fun changeMethodTargetToStatic(target: Tr.MethodInvocation, toClass: String) =
            changeMethodTargetToStatic(listOf(target), toClass)

    /**
     * Change to a static method invocation on <code>toClass</code>
     */
    fun changeMethodTargetToStatic(target: Tr.MethodInvocation, toClass: Class<*>) =
        changeMethodTargetToStatic(target, toClass.name)

    fun changeMethodTarget(targets: Iterable<Tr.MethodInvocation>, namedVar: String, type: Type.Class): Refactor {
        targets.forEach { target ->
            addOp(target, ChangeMethodTargetToVariable(namedVar, type))

            // if the original is a static method invocation, the import on it's type may no longer be needed
            target.type?.declaringType?.fullyQualifiedName?.let { addOp(RemoveImport(it)) }
        }

        return this
    }

    fun changeMethodTarget(target: Tr.MethodInvocation, namedVar: Tr.VariableDecls.NamedVar) =
            changeMethodTarget(target, namedVar.simpleName, namedVar.type.asClass()!!)

    fun changeMethodTarget(target: Tr.MethodInvocation, namedVar: String, type: Type.Class) =
            changeMethodTarget(listOf(target), namedVar, type)

    fun insertArgument(targets: Iterable<Tr.MethodInvocation>, pos: Int, source: String): Refactor {
        targets.forEach { target -> addOp(target, InsertMethodArgument(pos, source)) }
        return this
    }

    fun insertArgument(target: Tr.MethodInvocation, pos: Int, source: String) =
            insertArgument(listOf(target), pos, source)

    fun deleteArgument(target: Tr.MethodInvocation, pos: Int) =
            deleteArgument(listOf(target), pos)

    fun deleteArgument(targets: Iterable<Tr.MethodInvocation>, pos: Int): Refactor {
        targets.forEach { target -> addOp(target, DeleteMethodArgument(pos)) }
        return this
    }

    fun reorderArguments(target: Tr.MethodInvocation, vararg byArgumentNames: String): ReorderMethodArguments {
        val reorderOp = ReorderMethodArguments(byArgumentNames.toList())
        addOp(target, reorderOp)
        return reorderOp
    }

    // -------------
    // Expression Refactoring
    // -------------

    fun changeLiteral(targets: Iterable<Expression>, transform: (Any?) -> Any?): Refactor {
        targets.forEach { target -> addOp(target, ChangeLiteral(transform)) }
        return this
    }

    fun changeLiteral(target: Expression, transform: (Any?) -> Any?): Refactor =
            changeLiteral(listOf(target), transform)

    fun stats(): Map<String, Int> {
        val stats = mutableMapOf<String, Int>()

        ops.fold(original) { acc, (targetId, visitor) ->
            // by transforming the AST for each op, we allow for the possibility of overlapping changes
            val target = RetrieveTreeVisitor(targetId).visit(acc)
            val transformations = visitor.visit(target)
            val transformed = TransformVisitor(transformations).visit(acc) as Tr.CompilationUnit
            transformations.groupBy { it.name }.forEach { name, transformations ->
                stats.merge(name, transformations.size, Int::plus)
            }
            transformed
        }

        return stats
    }

    /**
     * @return Transformed version of the AST after changes are applied
     */
    fun fix(): Tr.CompilationUnit = ops
            .fold(original) { acc, (targetId, visitor) ->
                val target = RetrieveTreeVisitor(targetId).visit(acc)

                // by transforming the AST for each op, we allow for the possibility of overlapping changes
                TransformVisitor(visitor.visit(target)).visit(acc) as Tr.CompilationUnit
            }
            .let { cu -> TransformVisitor(FormatVisitor().visit(cu)).visit(cu) as Tr.CompilationUnit }

    /**
     * @return Git-style patch diff representing the changes to this compilation unit
     */
    @JvmOverloads
    fun diff(relativeTo: Path? = null) = InMemoryDiffEntry(Paths.get(original.sourcePath), relativeTo, original.print(), fix().print()).diff

    internal class InMemoryDiffEntry(filePath: Path, relativeTo: Path?, old: String, new: String): DiffEntry() {
        private val repo = InMemoryRepository.Builder().build()
        private val relativePath = relativeTo?.let { filePath.relativize(relativeTo) } ?: filePath

        init {
            changeType = ChangeType.MODIFY
            oldPath = relativePath.toString()
            newPath = relativePath.toString()

            val inserter = repo.objectDatabase.newInserter()
            oldId = inserter.insert(Constants.OBJ_BLOB, old.toByteArray()).abbreviate(40)
            newId = inserter.insert(Constants.OBJ_BLOB, new.toByteArray()).abbreviate(40)
            inserter.flush()

            oldMode = FileMode.REGULAR_FILE
            newMode = FileMode.REGULAR_FILE
            repo.close()
        }

        val diff: String by lazy {
            if(oldId == newId)
                ""
            else {
                val patch = ByteArrayOutputStream()
                val formatter = DiffFormatter(patch)
                formatter.setRepository(repo)
                formatter.format(this)
                String(patch.toByteArray())
            }
        }
    }
}